File: api/schemas/sources.py
"""
Pydantic schemas for Sources & Ingestion (Sprint 2).
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal
from pydantic import BaseModel, ConfigDict, Field

from api.schemas.status import DataMode

SourceDomain = Literal[
    "electoral", "legislative", "media", "economic",
    "regulatory", "geopolitical", "osint", "territorial",
    "contracts", "workspace", "system",
]

SourceStatus = Literal["active", "degraded", "down", "unknown", "disabled"]
SourceMode = Literal["api", "rss", "scraper", "file", "manual", "database", "pipeline"]


class SourceDefinition(BaseModel):
    model_config = ConfigDict(frozen=True)

    id: str
    name: str
    domain: SourceDomain
    mode: SourceMode
    description: str = ""
    url: str | None = None
    owner: str | None = None
    refresh_policy: str | None = None
    enabled: bool = True
    tags: list[str] = Field(default_factory=list)
    legal_notes: str | None = None


class SourceHealth(BaseModel):
    source_id: str
    status: SourceStatus
    last_success_at: datetime | None = None
    last_attempt_at: datetime | None = None
    last_error: str | None = None
    latency_ms: int | None = None
    records_last_run: int = 0
    records_24h: int = 0
    quality_score: float | None = None
    freshness_score: float | None = None
    coverage_score: float | None = None
    mode: DataMode = "real"


class SourceWithHealth(BaseModel):
    definition: SourceDefinition
    health: SourceHealth


class IngestionRunRequest(BaseModel):
    source_id: str
    dry_run: bool = True
    limit: int | None = 100
    force: bool = False


class IngestionRunResult(BaseModel):
    run_id: str
    source_id: str
    dry_run: bool
    status: Literal["queued", "running", "success", "warning", "error", "skipped"]
    started_at: datetime
    finished_at: datetime | None = None
    records_seen: int = 0
    records_new: int = 0
    records_updated: int = 0
    records_failed: int = 0
    message: str = ""
    error: str | None = None
    mode: DataMode = "real"

File: apps/web/lib/types/sources.ts
import type { DataMode } from "./status";

export type SourceDomain =
  | "electoral" | "legislative" | "media" | "economic"
  | "regulatory" | "geopolitical" | "osint" | "territorial"
  | "contracts" | "workspace" | "system";

export type SourceStatus = "active" | "degraded" | "down" | "unknown" | "disabled";
export type SourceIngestionMode = "api" | "rss" | "scraper" | "file" | "manual" | "database" | "pipeline";

export interface SourceDefinition {
  id: string;
  name: string;
  domain: SourceDomain;
  mode: SourceIngestionMode;
  description: string;
  url: string | null;
  owner: string | null;
  refresh_policy: string | null;
  enabled: boolean;
  tags: string[];
  legal_notes: string | null;
}

export interface SourceHealth {
  source_id: string;
  status: SourceStatus;
  last_success_at: string | null;
  last_attempt_at: string | null;
  last_error: string | null;
  latency_ms: number | null;
  records_last_run: number;
  records_24h: number;
  quality_score: number | null;
  freshness_score: number | null;
  coverage_score: number | null;
  mode: DataMode;
}

export interface SourceWithHealth {
  definition: SourceDefinition;
  health: SourceHealth;
}

export interface SourceHealthSummary {
  total: number;
  active: number;
  degraded: number;
  down: number;
  unknown: number;
  disabled: number;
}

export interface SourceCoverage {
  domain: SourceDomain;
  total: number;
  active: number;
  degraded: number;
  down: number;
  unknown: number;
  coverage_score?: number | null;
}

export interface IngestionRunRequest {
  source_id: string;
  dry_run: boolean;
  limit?: number | null;
  force?: boolean;
}

export interface IngestionRunResult {
  run_id: string;
  source_id: string;
  dry_run: boolean;
  status: "queued" | "running" | "success" | "warning" | "error" | "skipped";
  started_at: string;
  finished_at: string | null;
  records_seen: number;
  records_new: number;
  records_updated: number;
  records_failed: number;
  message: string;
  error: string | null;
  mode: DataMode;
}

// API response shapes
export interface SourcesCatalogResponse {
  mode: DataMode;
  sources: SourceDefinition[];
  total: number;
}

export interface SourcesHealthResponse {
  mode: DataMode;
  items: SourceWithHealth[];
  summary: SourceHealthSummary;
}

export interface SourcesCoverageResponse {
  mode: DataMode;
  domains: SourceCoverage[];
}

export interface SourcesRunsResponse {
  mode: DataMode;
  items: IngestionRunResult[];
  total: number;
}
